File: backend/app/api/v1/applications.py
from typing import Annotated
from uuid import UUID

from fastapi import (
    APIRouter,
    HTTPException,
    Query,
    Response,
    status,
)

from app.api.dependencies import (
    CurrentUser,
)
from app.api.dependencies.application_timeline import (
    ApplicationTimelineServiceDependency,
)
from app.api.dependencies.applications import (
    ApplicationServiceDependency,
)
from app.schemas.application import (
    ApplicationCreateRequest,
    ApplicationPageResponse,
    ApplicationResponse,
    ApplicationSortField,
    ApplicationStatus,
    ApplicationUpdateRequest,
    SortOrder,
)
from app.schemas.application_timeline import (
    ApplicationTimelineCreateRequest,
    ApplicationTimelineResponse,
    ApplicationTimelineUpdateRequest,
    TimelineSortOrder,
)
from app.services.application_timeline import (
    TimelineApplicationNotFoundError,
    TimelineEventNotFoundError,
    TimelinePersistenceError,
)
from app.services.applications import (
    ApplicationNotFoundError,
    ApplicationPersistenceError,
)

router = APIRouter(
    prefix="/applications",
    tags=["applications"],
)


@router.post(
    "",
    response_model=(ApplicationResponse),
    status_code=(status.HTTP_201_CREATED),
)
async def create_application(
    payload: ApplicationCreateRequest,
    current_user: CurrentUser,
    service: (ApplicationServiceDependency),
) -> ApplicationResponse:
    try:
        application = await service.create(
            user=current_user,
            payload=payload,
            source="manual",
        )
    except ApplicationPersistenceError as exc:
        raise HTTPException(
            status_code=(status.HTTP_500_INTERNAL_SERVER_ERROR),
            detail=str(exc),
        ) from exc

    return ApplicationResponse.model_validate(application)


@router.get(
    "",
    response_model=(ApplicationPageResponse),
)
async def list_applications(
    current_user: CurrentUser,
    service: (ApplicationServiceDependency),
    search: Annotated[
        str | None,
        Query(max_length=200),
    ] = None,
    status_filter: Annotated[
        ApplicationStatus | None,
        Query(alias="status"),
    ] = None,
    sort_by: Annotated[
        ApplicationSortField,
        Query(),
    ] = "applied_at",
    sort_order: Annotated[
        SortOrder,
        Query(),
    ] = "desc",
    page: Annotated[
        int,
        Query(ge=1),
    ] = 1,
    page_size: Annotated[
        int,
        Query(
            ge=1,
            le=100,
        ),
    ] = 20,
) -> ApplicationPageResponse:
    applications, total = await service.list_for_user(
        user=current_user,
        search=search,
        status=status_filter,
        sort_by=sort_by,
        sort_order=sort_order,
        page=page,
        page_size=page_size,
    )

    return ApplicationPageResponse(
        items=[ApplicationResponse.model_validate(application) for application in applications],
        total=total,
        page=page,
        page_size=page_size,
    )


@router.get(
    "/{application_id}",
    response_model=ApplicationResponse,
)
async def get_application(
    application_id: UUID,
    current_user: CurrentUser,
    service: ApplicationServiceDependency,
) -> ApplicationResponse:
    try:
        application = await service.get(
            user=current_user,
            application_id=application_id,
        )
    except ApplicationNotFoundError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(exc),
        ) from exc

    return ApplicationResponse.model_validate(application)


@router.patch(
    "/{application_id}",
    response_model=(ApplicationResponse),
)
async def update_application(
    application_id: UUID,
    payload: (ApplicationUpdateRequest),
    current_user: CurrentUser,
    service: (ApplicationServiceDependency),
) -> ApplicationResponse:
    try:
        application = await service.update(
            user=current_user,
            application_id=(application_id),
            payload=payload,
        )
    except ApplicationNotFoundError as exc:
        raise HTTPException(
            status_code=(status.HTTP_404_NOT_FOUND),
            detail=str(exc),
        ) from exc
    except ApplicationPersistenceError as exc:
        raise HTTPException(
            status_code=(status.HTTP_500_INTERNAL_SERVER_ERROR),
            detail=str(exc),
        ) from exc

    return ApplicationResponse.model_validate(application)


@router.delete(
    "/{application_id}",
    status_code=(status.HTTP_204_NO_CONTENT),
)
async def delete_application(
    application_id: UUID,
    current_user: CurrentUser,
    service: (ApplicationServiceDependency),
) -> Response:
    try:
        await service.delete(
            user=current_user,
            application_id=(application_id),
        )
    except ApplicationNotFoundError as exc:
        raise HTTPException(
            status_code=(status.HTTP_404_NOT_FOUND),
            detail=str(exc),
        ) from exc
    except ApplicationPersistenceError as exc:
        raise HTTPException(
            status_code=(status.HTTP_500_INTERNAL_SERVER_ERROR),
            detail=str(exc),
        ) from exc

    return Response(status_code=(status.HTTP_204_NO_CONTENT))


# DAY25_TIMELINE_ROUTES


@router.get(
    "/{application_id}/timeline",
    response_model=list[ApplicationTimelineResponse],
)
async def list_application_timeline(
    application_id: UUID,
    current_user: CurrentUser,
    service: (ApplicationTimelineServiceDependency),
    order: Annotated[
        TimelineSortOrder,
        Query(),
    ] = "asc",
) -> list[ApplicationTimelineResponse]:
    try:
        events = await service.list_events(
            application_id=application_id,
            user=current_user,
            order=order,
        )
    except TimelineApplicationNotFoundError as exc:
        raise HTTPException(
            status_code=(status.HTTP_404_NOT_FOUND),
            detail=str(exc),
        ) from exc

    return [ApplicationTimelineResponse.model_validate(event) for event in events]


@router.post(
    "/{application_id}/timeline",
    response_model=(ApplicationTimelineResponse),
    status_code=(status.HTTP_201_CREATED),
)
async def create_application_timeline_event(
    application_id: UUID,
    payload: (ApplicationTimelineCreateRequest),
    current_user: CurrentUser,
    service: (ApplicationTimelineServiceDependency),
) -> ApplicationTimelineResponse:
    try:
        event = await service.create(
            application_id=application_id,
            user=current_user,
            payload=payload,
            source="manual",
        )
    except TimelineApplicationNotFoundError as exc:
        raise HTTPException(
            status_code=(status.HTTP_404_NOT_FOUND),
            detail=str(exc),
        ) from exc
    except TimelinePersistenceError as exc:
        raise HTTPException(
            status_code=(status.HTTP_500_INTERNAL_SERVER_ERROR),
            detail=str(exc),
        ) from exc

    return ApplicationTimelineResponse.model_validate(event)


@router.patch(
    "/{application_id}/timeline/{event_id}",
    response_model=(ApplicationTimelineResponse),
)
async def update_application_timeline_event(
    application_id: UUID,
    event_id: UUID,
    payload: (ApplicationTimelineUpdateRequest),
    current_user: CurrentUser,
    service: (ApplicationTimelineServiceDependency),
) -> ApplicationTimelineResponse:
    try:
        event = await service.update(
            event_id=event_id,
            application_id=application_id,
            user=current_user,
            payload=payload,
        )
    except (
        TimelineApplicationNotFoundError,
        TimelineEventNotFoundError,
    ) as exc:
        raise HTTPException(
            status_code=(status.HTTP_404_NOT_FOUND),
            detail=str(exc),
        ) from exc
    except TimelinePersistenceError as exc:
        raise HTTPException(
            status_code=(status.HTTP_500_INTERNAL_SERVER_ERROR),
            detail=str(exc),
        ) from exc

    return ApplicationTimelineResponse.model_validate(event)


@router.delete(
    "/{application_id}/timeline/{event_id}",
    status_code=(status.HTTP_204_NO_CONTENT),
)
async def delete_application_timeline_event(
    application_id: UUID,
    event_id: UUID,
    current_user: CurrentUser,
    service: (ApplicationTimelineServiceDependency),
) -> Response:
    try:
        await service.delete(
            event_id=event_id,
            application_id=application_id,
            user=current_user,
        )
    except (
        TimelineApplicationNotFoundError,
        TimelineEventNotFoundError,
    ) as exc:
        raise HTTPException(
            status_code=(status.HTTP_404_NOT_FOUND),
            detail=str(exc),
        ) from exc
    except TimelinePersistenceError as exc:
        raise HTTPException(
            status_code=(status.HTTP_500_INTERNAL_SERVER_ERROR),
            detail=str(exc),
        ) from exc

    return Response(status_code=(status.HTTP_204_NO_CONTENT))

File: frontend/src/lib/applications.ts
import { clearAuthSession, loadAuthSession } from "@/lib/auth-storage";
import { env } from "@/lib/env";

export type ApplicationStatus =
  "applied" | "online_assessment" | "interview" | "offer" | "rejected" | "withdrawn";

export type ApplicationSource = "manual" | "gmail" | "integration";

export type ApplicationSortField = "applied_at" | "created_at" | "company" | "role" | "status";

export interface Application {
  id: string;
  company: string;
  role: string;
  job_url: string | null;
  location: string | null;
  applied_at: string;
  status: ApplicationStatus;
  source: ApplicationSource;
  external_id: string | null;
  notes: string | null;
  created_at: string;
  updated_at: string;
}

export interface ApplicationCreateInput {
  company: string;
  role: string;
  job_url?: string | null;
  location?: string | null;
  applied_at: string;
  status: ApplicationStatus;
  notes?: string | null;
}

export interface ApplicationPage {
  items: Application[];
  total: number;
  page: number;
  page_size: number;
}

interface ListApplicationOptions {
  search?: string;
  status?: ApplicationStatus | "";
  sortBy?: ApplicationSortField;
  sortOrder?: "asc" | "desc";
  page?: number;
  pageSize?: number;
}

function getErrorMessage(body: unknown, fallback: string): string {
  if (typeof body === "object" && body !== null && "detail" in body) {
    const detail = (
      body as {
        detail?: unknown;
      }
    ).detail;

    if (typeof detail === "string") {
      return detail;
    }
  }

  return fallback;
}

async function authenticatedRequest<T>(path: string, init: RequestInit = {}): Promise<T> {
  const session = loadAuthSession();

  if (!session) {
    throw new Error("Your session has expired. Please sign in again.");
  }

  let response: Response;

  try {
    response = await fetch(`${env.apiBaseUrl}${path}`, {
      ...init,
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${session.accessToken}`,
        ...init.headers,
      },
    });
  } catch {
    throw new Error("Unable to reach the backend API. Make sure FastAPI is running on port 8000.");
  }

  if (response.status === 401) {
    clearAuthSession();

    if (typeof window !== "undefined") {
      window.location.assign("/");
    }

    throw new Error("Your session has expired. Please sign in again.");
  }

  if (response.status === 204) {
    return undefined as T;
  }

  const body: unknown = await response.json().catch(() => null);

  if (!response.ok) {
    throw new Error(getErrorMessage(body, `Request failed with status ${response.status}.`));
  }

  return body as T;
}

export async function listApplications(
  options: ListApplicationOptions = {},
): Promise<ApplicationPage> {
  const params = new URLSearchParams();

  if (options.search?.trim()) {
    params.set("search", options.search.trim());
  }

  if (options.status) {
    params.set("status", options.status);
  }

  params.set("sort_by", options.sortBy ?? "applied_at");

  params.set("sort_order", options.sortOrder ?? "desc");

  params.set("page", String(options.page ?? 1));

  params.set("page_size", String(options.pageSize ?? 20));

  return authenticatedRequest<ApplicationPage>(`/applications?${params.toString()}`);
}

export async function createApplication(payload: ApplicationCreateInput): Promise<Application> {
  return authenticatedRequest<Application>("/applications", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export async function getApplication(applicationId: string): Promise<Application> {
  return authenticatedRequest<Application>(`/applications/${applicationId}`);
}

export async function updateApplication(
  applicationId: string,
  payload: Partial<ApplicationCreateInput>,
): Promise<Application> {
  return authenticatedRequest<Application>(`/applications/${applicationId}`, {
    method: "PATCH",
    body: JSON.stringify(payload),
  });
}

export async function deleteApplication(applicationId: string): Promise<void> {
  await authenticatedRequest<void>(`/applications/${applicationId}`, {
    method: "DELETE",
  });
}

export type TimelineEventType =
  | "application_submitted"
  | "status_changed"
  | "online_assessment_received"
  | "online_assessment_completed"
  | "interview_scheduled"
  | "interview_completed"
  | "offer_received"
  | "rejected"
  | "withdrawn"
  | "note";

export type TimelineEventSource = "manual" | "system" | "gmail" | "integration";

export interface ApplicationTimelineEvent {
  id: string;
  application_id: string;
  event_type: TimelineEventType;
  title: string;
  description: string | null;
  related_status: ApplicationStatus | null;
  source: TimelineEventSource;
  external_id: string | null;
  event_at: string;
  created_at: string;
  updated_at: string;
}

export interface TimelineEventCreateInput {
  event_type: TimelineEventType;
  title: string;
  description?: string | null;
  related_status?: ApplicationStatus | null;
  event_at: string;
}

export interface TimelineEventUpdateInput {
  event_type?: TimelineEventType;
  title?: string;
  description?: string | null;
  related_status?: ApplicationStatus | null;
  event_at?: string;
}

export async function listApplicationTimeline(
  applicationId: string,
): Promise<ApplicationTimelineEvent[]> {
  return authenticatedRequest<ApplicationTimelineEvent[]>(
    `/applications/${applicationId}/timeline?order=asc`,
  );
}

export async function createApplicationTimelineEvent(
  applicationId: string,
  payload: TimelineEventCreateInput,
): Promise<ApplicationTimelineEvent> {
  return authenticatedRequest<ApplicationTimelineEvent>(`/applications/${applicationId}/timeline`, {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export async function updateApplicationTimelineEvent(
  applicationId: string,
  eventId: string,
  payload: TimelineEventUpdateInput,
): Promise<ApplicationTimelineEvent> {
  return authenticatedRequest<ApplicationTimelineEvent>(
    `/applications/${applicationId}/timeline/${eventId}`,
    {
      method: "PATCH",
      body: JSON.stringify(payload),
    },
  );
}

export async function deleteApplicationTimelineEvent(
  applicationId: string,
  eventId: string,
): Promise<void> {
  await authenticatedRequest<void>(`/applications/${applicationId}/timeline/${eventId}`, {
    method: "DELETE",
  });
}
